### PR TITLE
[ML-11010] add a search registered models database query and tests

### PR DESCRIPTION
**Pull request by ankit-db**
## What changes are proposed in this pull request?

This PR introduces a database query that can search for registered models via an equality operator, `LIKE`, and `ILIKE`. This is going to eventually be a part of a new API introduced to search for registered models.

## How is this patch tested?

Unit tests

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [x] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for
Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [x] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/store/model_registry/sqlalchemy_store.py [hunk 1]
```diff
@@ -205,6 +205,48 @@ def list_registered_models(self):
...
+            else:
+                raise MlflowException('Invalid filter string: %s' % filter_string,
+                                      error_code=INVALID_PARAMETER_VALUE)
+        else:
+            raise MlflowException('Model Registry expects filter to be like '
+                                  '"name (=)(LIKE)(ILIKE) \'<model_name>\'"'
+                                  'Input filter string: %s. ' % filter_string,
+                                  error_code=INVALID_PARAMETER_VALUE)
+        with self.ManagedSessionMaker() as session:
+            sql_registered_models = session.query(SqlRegisteredModel).filter(*conditions).all()
```
Comment (mparkhe) [inline mlflow/store/model_registry/sqlalchemy_store.py:246]: you may need to add some logic here to handle case sensitivity.

File: mlflow/store/model_registry/sqlalchemy_store.py [hunk 2]
```diff
@@ -205,6 +205,48 @@ def list_registered_models(self):
...
+                    conditions = [SqlRegisteredModel.name.like(filter_dict["value"])]
+                elif filter_dict["comparator"] == "ILIKE":
+                    conditions = [SqlRegisteredModel.name.ilike(filter_dict["value"])]
+                else:
+                    conditions = [SqlRegisteredModel.name == filter_dict["value"]]
+            else:
+                raise MlflowException('Invalid filter string: %s' % filter_string,
+                                      error_code=INVALID_PARAMETER_VALUE)
+        else:
+            raise MlflowException('Model Registry expects filter to be like '
```
Comment (mparkhe) [inline mlflow/store/model_registry/sqlalchemy_store.py:241]: Change message to something like --

> Invalid filter string '%s'. Search registered models supports filter expression of the type "name <op> 'model-name'". Supported <op> are equality (=) comparator...

(you can reuse that part of the message from the above error message. I would create a local variable so you do not need to change two places when we add more operator support)

File: tests/store/model_registry/test_sqlalchemy_store.py [hunk 3]
```diff
@@ -381,3 +381,100 @@ def search_versions(filter_string):
...
+        self.assertEqual(set(search_registered_model(f"name LIKE '%RM%'")),
+                         set([name1, name2, name3, name4, name5, name6]))
+
+        # case-sensitive prefix search using LIKE with surrounding % should return all the RMs
+        self.assertEqual(set(search_registered_model(f"name LIKE '_e%'")),
+                         set([name1, name2, name3, name4, name5, name6]))
+
+        # case-sensitive prefix search using LIKE should return just rm5 and rm6
+        # note that this test case returns both because the test DB is in sqlite. In sqlite,
+        # LIKE is NOT case-sensitive.
```
[the remark this answers is not in the document]
Comment (mparkhe) [reply tests/store/model_registry/test_sqlalchemy_store.py:428]: found it:
https://github.com/mlflow/mlflow/blob/master/mlflow/store/tracking/sqlalchemy_store.py#L627

File: mlflow/store/model_registry/sqlalchemy_store.py (hunk 2 above)
Comment (mparkhe) [reply mlflow/store/model_registry/sqlalchemy_store.py:241]: cc: @andreakress

File: tests/store/model_registry/test_sqlalchemy_store.py [hunk 4]
```diff
@@ -381,3 +381,100 @@ def search_versions(filter_string):
...
+        self._rm_maker(name3)
+        self._rm_maker(name4)
+        self._rm_maker(name5)
+        self._rm_maker(name6)
+
+        def search_registered_model(filter_string):
+            return [rm.name for rm in self.store.search_registered_models(filter_string)]
+
+        # search with no filter should return all registered models
+        self.assertEqual(set(search_registered_model("")),
```
[the remark this answers is not in the document]
Comment (ankitmathur-db) [reply tests/store/model_registry/test_sqlalchemy_store.py:405]: It is treated as such by the model registry filter logic from search for model versions - that code accepts None or empty string

File: mlflow/store/model_registry/sqlalchemy_store.py [hunk 5]
```diff
@@ -205,6 +205,48 @@ def list_registered_models(self):
...
+                 objects.
+        """
+        parsed_filter = SearchUtils.parse_filter_for_model_registry(filter_string)
+        if len(parsed_filter) == 0:
+            conditions = []
+        elif len(parsed_filter) == 1:
+            filter_dict = parsed_filter[0]
+            if filter_dict["comparator"] not in \
+                    SearchUtils.VALID_REGISTERED_MODEL_SEARCH_COMPARATORS:
+                raise MlflowException('Model Registry search filter only supports equality(=) '
```
[the remark this answers is not in the document]
Comment (ankitmathur-db) [reply mlflow/store/model_registry/sqlalchemy_store.py:225]: Yikes. Shockingly unreadable on my part - adjusting the language

File: mlflow/store/model_registry/sqlalchemy_store.py (hunk 1 above)
Comment (ankitmathur-db) [reply mlflow/store/model_registry/sqlalchemy_store.py:246]: Done

File: tests/store/model_registry/test_sqlalchemy_store.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ankitmathur-db) [reply tests/store/model_registry/test_sqlalchemy_store.py:428]: Making it case sensitive is definitely the move if we can achieve that experience across the board. The user shouldn't have to remember this particular idiosyncrasy - adjusted the code in the search function to account for this.

File: mlflow/store/model_registry/sqlalchemy_store.py [hunk 6]
```diff
@@ -205,6 +206,51 @@ def list_registered_models(self):
...
+                    conditions = [SqlRegisteredModel.name.ilike(filter_dict["value"])]
+                else:
+                    conditions = [SqlRegisteredModel.name == filter_dict["value"]]
+            else:
+                raise MlflowException('Invalid filter string: %s' % filter_string,
+                                      error_code=INVALID_PARAMETER_VALUE)
+        else:
+            supported_ops = f'"name ' \
+                f'{"/".join(SearchUtils.VALID_REGISTERED_MODEL_SEARCH_COMPARATORS)} ' \
+                f'\'<model_name>\''
```
Comment (mparkhe) [inline mlflow/store/model_registry/sqlalchemy_store.py:243]: Acceptable filter message will look something like this -- 

`"name" =/LIKE/ILIKE 'model_name'"`

Super confusing. The way you had it before was simpler to understand.

File: tests/store/model_registry/test_sqlalchemy_store.py [hunk 7]
```diff
@@ -381,3 +381,100 @@ def search_versions(filter_string):
...
+        # case-sensitive prefix search using LIKE should return all the RMs
+        self.assertEqual(set(search_registered_model(f"name LIKE '{prefix}%'")),
+                         set([name1, name2, name3, name4, name5, name6]))
+
+        # case-sensitive prefix search using LIKE with surrounding % should return all the RMs
+        self.assertEqual(set(search_registered_model(f"name LIKE '%RM%'")),
+                         set([name1, name2, name3, name4, name5, name6]))
+
+        # case-sensitive prefix search using LIKE with surrounding % should return all the RMs
+        self.assertEqual(set(search_registered_model(f"name LIKE '_e%'")),
```
Comment (mparkhe) [inline tests/store/model_registry/test_sqlalchemy_store.py:423]: Please put a comment that `"_e%"` matched `"test_fo.."`

File: tests/store/model_registry/test_sqlalchemy_store.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mparkhe) [reply tests/store/model_registry/test_sqlalchemy_store.py:428]: `LIKE` should be case-sensitive everywhere and with your `PRAGMA` fix SQLite should also be fixed.

Maybe this comment above this line of code needs to be removed.

**Review by mparkhe — COMMENTED** [review]
Looks good.
The error message is confusing and couple of **nits**

File: mlflow/store/model_registry/sqlalchemy_store.py (hunk 6 above)
Comment (ankitmathur-db) [reply mlflow/store/model_registry/sqlalchemy_store.py:243]: The way I had it before was something like
```
"name" (=)(LIKE)(ILIKE) 'model_name'"
```
Do you think that looks better?

Comment (andreakress) [reply mlflow/store/model_registry/sqlalchemy_store.py]: I think this is a good, clear error message.

**Review by mparkhe — APPROVED** [review]

File: mlflow/store/model_registry/sqlalchemy_store.py (hunk 6 above)
Comment (ankitmathur-db) [reply mlflow/store/model_registry/sqlalchemy_store.py]: Thanks for your feedback @andreakress!

**Comment by ankitmathur-db** [comment]
Merging, per our offline discussion!

**Pull request merged**